File: edp_java/src/main/Counter.java
package main;

import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

public class Counter {
    private static final BlockingQueue<Order> orders = new LinkedBlockingQueue<>();
    private static final BlockingQueue<Order> completedMeals = new LinkedBlockingQueue<>();

    private Counter() {
    }

    public static void placeOrder(Order order) {
        try {
            orders.put(order);
            System.out.println(String.format("Order for %s placed on counter.", order.philosopher().name()));
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    public static Order takeOrder() {
        try {
            Order order = orders.take();
            System.out
                    .println(String.format("Order for %s picked up from counter.", order.philosopher().name()));
            return order;
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            return null;
        }
    }

    public static boolean hasOrders() {
        return !orders.isEmpty();
    }

    public static int orderCount() {
        return orders.size();
    }

    public static void placeCompletedMeal(Order order) {
        try {
            completedMeals.put(order);
            System.out.println(String.format("Order for %s placed on counter.", order.philosopher().name()));
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    public static boolean hasCompletedMeals() {
        return !completedMeals.isEmpty();
    }

    public static Order pollCompletedMeal() {
        Order order = completedMeals.poll();
        if (order != null) {
            System.out
                    .println(String.format("Order for %s picked up from counter.", order.philosopher().name()));
        }
        return order;
    }
}

File: edp_go/counter.go
package main

import "sync"

type Counter struct {
	orders          chan *Order
	completedMeals  chan *Order
	orderCount      int
	orderCountMutex sync.RWMutex
}

var counter = &Counter{
	orders:         make(chan *Order, 100),
	completedMeals: make(chan *Order, 100),
}

func (c *Counter) PlaceOrder(order *Order) {
	c.orders <- order
	c.orderCountMutex.Lock()
	c.orderCount++
	c.orderCountMutex.Unlock()
}

func (c *Counter) TakeOrder() *Order {
	select {
	case order := <-c.orders:
		c.orderCountMutex.Lock()
		c.orderCount--
		c.orderCountMutex.Unlock()
		return order
	default:
		return nil
	}
}

func (c *Counter) HasOrders() bool { return len(c.orders) > 0 }

func (c *Counter) OrderCount() int {
	c.orderCountMutex.RLock()
	defer c.orderCountMutex.RUnlock()
	return c.orderCount
}

func (c *Counter) PlaceCompletedMeal(order *Order) { c.completedMeals <- order }

func (c *Counter) HasCompletedMeals() bool { return len(c.completedMeals) > 0 }

func (c *Counter) PollCompletedMeal() *Order {
	select {
	case order := <-c.completedMeals:
		return order
	default:
		return nil
	}
} 